### ISSUE TITLE
Japanese character conversion *pretty much* unusable on macOS

### ISSUE THREAD

**Issue by iorippi** [randy3k/Terminus#104]
Although there *are* a work around (written at the very end), there is a huge issue typing and converting Japanese texts in this.

I have tested on multiple OS-es, and ALL problems that I mentioned here were exclusively happening on macOS. And they all share the list of plugins, even the configs besides Terminus' default shell configurations. (So there were no hassle at all on Linux/Win10 anywhere.)

- macOS (Mojave, English / Japanese) BASH, FISH
- Linux (Fedora 29, English / Japanese) BASH, FISH
- Windows (10, Japanese) CMD, BASH(WSL), FISH(WSL)

However, I have only tested Japanese so I'm not sure if this also is the case for Chinese/Korean. Anyone using those languages on macOS here?

---

![SCR-1 mov](https://user-images.githubusercontent.com/10939446/55888891-26541e80-5beb-11e9-9042-a146551b5a5a.gif)

Below is the close lookup.  I tried to be as concise as possible, but I think I just don't know how to do so.. This is like 4 issues put in one place so probably I should've split them, but I thought this makes more sense to put all in one, so here it is.
**Bold texts** are the ones with problems, but some may not use 1 or 2 much. Although 0 and 6 will be naughty pain for most of Japanese (and possibly Chinese/Korean?) typists.

0. **Retype text** (Such as deleting mistyped words and retype)  
    - <kbd>Delete</kbd>  
    Deletes the whole input characters instead of one by one
1. **Setting conversion range** (When default conversion range isn't desirable)  
    - <kbd>Shift</kbd> + <kbd>Left / Right</kbd>  
    directly casts `;2D` or `;2C` to console in some weird way  
2. **Select conversion range** (if there were multiple ranges to choose from)  
    - <kbd>Left / Right</kbd>  
      Entirely cancels the Japanese input operation, and at the same time directly casts `;2D` or `;2C` to console
3. Open conversion candidate list  
  (Obviously, it won't reach here if 1. or 2. was executed)
    - <kbd>Space</kbd>  
    Works as expected (Conversion candidate list gets displayed)
4. **Go through conversion candidate from the list**  
    - <kbd>Space</kbd>  
    Works as expected (Selected conversion will be visible in line)
    - <kbd>Up / Down</kbd>  
    Entirely cancels the Japanese input operation, and at the same time directly casts `;2D` or `;2C` to console.
5. Confirm a candidate from the list  
    - <kbd>Space</kbd>  
    Works as expected (Selected conversion will be visible in line)
    - <kbd>0-9</kbd>
    Works as expected (Selected conversion will be visible in line + moves onto the next conversion range)
6. **Finish the conversion and output the result**  
    - <kbd>Return</kbd>  
    Casts `return` to the console without conversion output  
    - <kbd>Shift</kbd> + <kbd>Return</kbd>  
    **Works !**

*(I purposely didn't test conversion using mouse/touchpad because it's not optimized for changing conversion range in the first place at the OS level.
Sometimes the input was displayed on the left end of the console lines over the position where hostname should be, which was too weird and I was honestly too tired to track down how to reproduce such thing, and so I'm not reporting that here neither.)*

---

Given that the configuration is exactly the same, the key bindings shouldn't be a problem here? But what else would be? 
This plugin would be a killer for my lazy-ass team to work with terminal so I'll do pretty much what I can do to find solution. (only if I was skilled enough to contrubute/debug..)

Basically, it does have workaround (Such as type-convert in small chunks of a word by word), but it's going to be super tedious for normal sentences, plus it will require some work to get used to unusual key combination, so viability is beyond question as of now.

**Comment by randy3k** [comment]
tldr; I can type Chinese on macOS. To investigate, I need to learn how to type Japanese on macOS (and I don't know Japanese).

**Comment by iorippi** [comment]
@randy3k I was hoping the whole scheme is shared between all CJK language (as it seemed to be work in similar manner), but I definitely could see that coming haha. Anyhow, thanks for checking!

Hopefully I can come back when I become able to debug by myself, or some fellows will do instead..

One last question: If Chinese input utilizes <kbd>Return</kbd> and <kbd>Arrows</kbd> in the process of conversion (on macOS), do they work as expected?

**Comment by randy3k** [comment]
Oops, I wrongly editted your post.

**Comment by randy3k** [comment]
0. **Retype text** (Such as deleting mistyped words and retype)  
    - same issue happens in Chinese

1. **Setting conversion range** (When default conversion range isn't desirable)  
    - we don't have this concept in Chinese, need investigation

2. **Select conversion range** (if there were multiple ranges to choose from)  
    - we don't have this concept in Chinese, need investigation

4. **Go through conversion candidate from the list**  
    - <kbd>Up / Down</kbd>  
    same issue happens in Chinese

6. **Finish the conversion and output the result**  
    - we don't have this concept in Chinese, need investigation

**Comment by iorippi** [comment]
> Oops, I wrongly editted your post.
@randy3k It's all good sir, it was too damn long and I didn't know how to summarize: Thanks for that!

**Comment by randy3k** [comment]
looking for a way to revert my changes to your post....(stupid me)

**Comment by randy3k** [comment]
Ok, I managed to save most of the post.

**Comment by iorippi** [comment]
@randy3k I can reformat for you, or maybe "delete revision from this post" will do?? Thanks for works!

**Comment by randy3k** [comment]
> @randy3k I was hoping the whole scheme is shared between all the CJK languages (as it seemed to be working in the similar manner), but I definitely could see that coming haha. Anyhow, thanks for checking!
> 
> Hopefully I can come back when I become able to debug by myself, or some fellows will do instead..
> 
> One last question: If Chinese input utilizes Return and Arrows in the process of conversion (on macOS), do they work as expected?


On macOS, it seems that when we define the keys such as `left`, `right`, `up` and `down`. Those keys can no longer interact with the input overlay.

In chinese typing (at least for the input method ChangJie), <kbd>shift+arrow</kbd> is the same as <kbd>arrow</kbd>

**Comment by iorippi** [comment]
@randy3k 
> In chinese typing (at least for the input method ChangJie), shift+arrow is the same as arrow and shift+return is the same as return. Maybe because we always convert.

That's interesting. And this is too confusing (as I can't see much logical consistency from my table below). 

*Key combination while in conversion process*

| Key combination | Chinese<br>(Terminus) | Japanese<br>(Terminus) | Japanese<br>(macOS terminal) |
| -- | -- | -- | -- |
| <kbd>return</kbd>           | works   | won't work (bypasses input overlay) | works |
| <kbd>shift+return</kbd> | functions like <kbd>return</kbd> | works | works |
| <kbd>arrows</kbd>          | works | won't work (bypasses input overlay) | works |
| <kbd>shift+arrows</kbd> | functions like <kbd>arrows</kbd> | won't work <br>(bypasses input overlay and outputs something like [;2A\|;2B\|;2C\|;2D] for [U\|D\|L\|R], similar to how macOS terminal behaves normally outside conversion process) | works

Thanks again for insight!

**Comment by randy3k** [comment]
For Chinese, arrows actually do not work. (I have updated the table).
Anyway, thanks for the nice summary.
